### PR TITLE
[wasm][debugger] Support display of DateTime, DateTimeOffset, Guid

### PR DESCRIPTION
**Pull request by kjpou1**
Support displaying the value types DateTime, DateTimeOffset and Guid values in the locals.

image

### CODE REVIEW

File: mono/mini/mini-wasm-debugger.c [hunk 1]
```diff
@@ -588,6 +597,16 @@ typedef struct {
 static gboolean describe_value(MonoType * type, gpointer addr)
 {
 	ERROR_DECL (error);
+
+	if (!datetime_class) {
+		datetime_class = mono_class_from_name_checked (mono_get_corlib(), "System", "DateTime", error);
+	}
```
Comment (vargaz) [inline mono/mini/mini-wasm-debugger.c]: These should use the GENERATE_TRY_GET_CLASS_WITH_CACHE macro:

static GENERATE_TRY_GET_CLASS_WITH_CACHE (fixed_buffer, "System.Runtime.CompilerServices", "FixedBufferAttribute")

File: mono/mini/mini-wasm-debugger.c [hunk 2]
```diff
@@ -662,6 +681,49 @@ static gboolean describe_value(MonoType * type, gpointer addr)
...
+					g_hash_table_insert_replace (val_to_types, GINT_TO_POINTER(objectid), GINT_TO_POINTER(1), TRUE);
+				}
+			} else if (klass == guid_class) {
+				MonoObject *obj = *(MonoObject**)addr;
+				if (!obj) {
+					mono_wasm_add_string_var (NULL);
+				} else {
+					int objectid = get_object_id(obj);
+					mono_wasm_add_obj_var ("System.Guid", objectid);
+					g_hash_table_insert_replace (obj_to_val_refs, GINT_TO_POINTER(objectid), addr, TRUE);
```
Comment (vargaz) [inline mono/mini/mini-wasm-debugger.c]: addr is a temporary value which is only valid as long as the debuggee is suspended. For this reason, it shouldn't be stored in hash tables etc. Instead, the whole valuetype is usually sent to the IDE in the normal non-wasm debugger.

Comment (kjpou1) [reply mono/mini/mini-wasm-debugger.c]: This has proven to be a problem so far.  Saw this in the debugger agent but so far the interaction messages are different.  Will look into this more.

File: mono/mini/mini-wasm-debugger.c (hunk 1 above)
Comment (kjpou1) [reply mono/mini/mini-wasm-debugger.c]: update in recent commits

**Comment by kjpou1** [comment]
@monojenkins build Linux WebAssembly

File: mono/mini/mini-wasm-debugger.c [hunk 3]
```diff
@@ -662,6 +674,49 @@ static gboolean describe_value(MonoType * type, gpointer addr)
 			}
 			break;
 		}
+		case MONO_TYPE_VALUETYPE:
+		case MONO_TYPE_TYPEDBYREF: {
+			MonoClass *klass = mono_class_from_mono_type_internal (type);
+			if (klass == mono_class_try_get_datetime_class ()) {
```
Comment (lewing) [inline mono/mini/mini-wasm-debugger.c]: It looks like these 3 blocks could be reduced to one that gets the the name from the klass? or at least reduce the duplication

File: mono/mini/mini-wasm-debugger.c [hunk 4]
```diff
@@ -683,44 +738,80 @@ describe_object_properties (guint64 objectId, gboolean isAsyncLocalThis)
...
 			mono_wasm_add_properties_var(p->name); 
 			sig = mono_method_signature_internal (p->get);
-			res = mono_runtime_try_invoke (p->get, obj, NULL, &exc, error);
+			if (!obj_val_ref)
+				res = mono_runtime_try_invoke (p->get, obj, NULL, &exc, error);
+			else
+			{
+				res = mono_runtime_try_invoke (p->get, (gpointer) obj_val_ref , NULL, &exc, error);
+			}
+
```
Comment (lewing) [inline mono/mini/mini-wasm-debugger.c]: reduce this to 
```
res = mono_runtime_try_invoke (p->get, obj_val_ref ? obj_val_ref : obj, NULL, &exc, error);
```

File: mono/mini/mini-wasm-debugger.c (hunk 3 above)
Comment (kjpou1) [reply mono/mini/mini-wasm-debugger.c]: In the works

**Comment by lewing** [comment]
part of #17876

File: mono/mini/mini-wasm-debugger.c (hunk 3 above)
Comment (kjpou1) [reply mono/mini/mini-wasm-debugger.c]: Addressed in recent commits.

**Comment by radical** [comment]
Closing as this was fixed via other PRs.

**Pull request closed without merging**